File: backend/utils/notifier.js
const db = require("../config/db");

/**
 * Inserts a notification record into the database for a specific user role or email.
 */
const createNotification = (email, type, priority, title, message) => {
    const sql = `
        INSERT INTO notifications (user_email, type, priority, title, message)
        VALUES (?, ?, ?, ?, ?)
    `;
    db.query(sql, [email || "system@hiring.com", type, priority || "NORMAL", title, message], (err) => {
        if (err) console.error("Failed to create backend notification:", err);
    });
};

/**
 * Broadcasts a notification to all HR and Admin users (recruiters) in a specific organization.
 */
const notifyRecruiters = (organizationId, type, priority, title, message) => {
    if (!organizationId) {
        console.warn("notifyRecruiters called without organizationId");
        return;
    }
    const sql = `
        SELECT u.email FROM users u
        INNER JOIN memberships m ON u.id = m.user_id
        WHERE m.organization_id = ?
          AND m.status = 'ACTIVE'
          AND (m.role IN ('HR', 'Admin') OR u.role IN ('HR', 'Admin'))
    `;
    db.query(sql, [organizationId], (err, rows) => {
        if (err || !rows) return;
        rows.forEach(user => {
            createNotification(user.email, type, priority, title, message);
        });
    });
};

module.exports = {
    createNotification,
    notifyRecruiters
};

File: backend/src/main/java/com/ats/service/Notifier.java
package com.ats.service;

import com.ats.entity.Notification;
import com.ats.entity.Membership;
import com.ats.entity.User;
import com.ats.repository.NotificationRepository;
import com.ats.repository.UserRepository;
import com.ats.repository.MembershipRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;

@Service
@Transactional
public class Notifier {

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private MembershipRepository membershipRepository;

    /**
     * Inserts a notification record into the database for a specific user email.
     */
    public void createNotification(String email, String type, String priority, String title, String message) {
        try {
            Notification notification = new Notification();
            notification.setUserEmail(email != null ? email : "system@hiring.com");
            notification.setType(type);
            notification.setPriority(priority != null ? priority : "NORMAL");
            notification.setTitle(title);
            notification.setMessage(message);
            notification.setIsRead(false);

            notificationRepository.save(notification);
        } catch (Exception e) {
            System.err.println("Failed to create backend notification: " + e.getMessage());
        }
    }

    /**
     * Broadcasts a notification to all HR and Admin users (recruiters) in a specific organization.
     */
    public void notifyRecruiters(Integer orgId, String type, String priority, String title, String message) {
        try {
            if (orgId == null) return;
            List<Membership> memberships = membershipRepository.findActiveHRAndAdmins(orgId);
            for (Membership m : memberships) {
                createNotification(m.getUser().getEmail(), type, priority, title, message);
            }
        } catch (Exception e) {
            System.err.println("Failed to notify recruiters: " + e.getMessage());
        }
    }
}
